packages/xod-client/src/hinting/validators.internal.js

```diff
@@ -0,0 +1,486 @@
+// NodeValidateFn :: Patch -> Project -> Map NodeId NodeErrors
+// LinkValidateFn :: Link -> Patch -> Project -> Map PatchPath DeducedPinTypes -> [Error]
+
+// ErrorsUpdateData :: { policy: UPDATE_ERRORS_POLICY, errors: Map PatchPath PatchErrors }
+
+const UPDATE_ERRORS_POLICY = {
+  OVERWRITE: 'OVERWRITE_ERRORS', // Overwrites whole object with new one
+  MERGE: 'MERGE_ERRORS', // Merges deeply new object into previvous one
+  ASSOC: 'ASSOC_ERRORS', // Overwrites only listed patch paths
+};
+
+// =============================================================================
+//
+// Basic validate functions
+//
+// =============================================================================
+
+export const getMarkerNodesErrorMap = (
+  predicate,
+  validator,
+  errorType
+) => patch => {
+  const markerNodeIds = R.compose(
+    R.map(XP.getNodeId),
+    R.filter(predicate),
+    XP.listNodes
+  )(patch);
+
+  if (R.isEmpty(markerNodeIds)) return {};
+
+  return foldEither(
+    err =>
+      R.compose(
+        mergeAllWithConcat,
+        R.map(R.objOf(R.__, { [errorType]: [err] }))
+      )(markerNodeIds),
+    R.always({ [errorType]: [] }),
+    validator(patch)
+  );
+};
+
+// :: Patch -> Map NodeId (Map ErrorType [Error])
+export const getVariadicMarkersErrorMap = getMarkerNodesErrorMap(
+  R.pipe(XP.getNodeType, XP.isVariadicPath),
+  XP.validatePatchForVariadics,
+  'validatePatchForVariadics'
+);
+
+// :: Patch -> Map NodeId (Map ErrorType [Error])
+export const getAbstractMarkersErrorMap = getMarkerNodesErrorMap(
+  R.pipe(XP.getNodeType, R.equals(XP.ABSTRACT_MARKER_PATH)),
+  XP.validateAbstractPatch,
+  'validateAbstractPatch'
+);
+
+// :: Patch -> Map NodeId (Map ErrorType [Error])
+export const getConstructorMarkersErrorMap = getMarkerNodesErrorMap(
+  R.pipe(XP.getNodeType, R.equals(XP.OUTPUT_SELF_PATH)),
+  XP.validateConstructorPatch,
+  'validateConstructorPatch'
+);
+
+// :: Patch -> Map NodeId (Map ErrorType [Error])
+export const getBusesErrorMap = R.compose(
+  R.map(R.objOf('validateBuses')),
+  foldEither(
+    err =>
+      R.compose(
+        R.map(R.of),
+        R.fromPairs,
+        R.map(R.pair(R.__, err)),
+        R.path(['payload', 'nodeIds'])
+      )(err),
+    R.always({})
+  ),
+  XP.validateBuses
+);
+
+// :: Patch -> Map NodeId (Map ErrorType [Error])
+export const getTerminalsErrorMap = R.compose(
+  R.map(R.objOf('validatePinLabels')),
+  foldEither(
+    err =>
+      R.compose(
+        R.map(R.of),
+        R.fromPairs,
+        R.map(R.pair(R.__, err)),
+        R.path(['payload', 'pinKeys']) // those are affected terminal node ids
+      )(err),
+    R.always([])
+  ),
+  XP.validatePinLabels
+);
+
+// TODO: Use validator from xod-project after refactoring
+// :: Patch -> Project -> Map NodeId (Map ErrorType [Error])
+export const getDeadRefErrorMap = (patch, project) =>
+  R.compose(
+    R.map(
+      R.compose(
+        R.objOf('checkPatchExists'),
```

Comment:
There are quite a lot of "magic" strings like this. Maybe they should be extracted to some constant?
Or, since most of them are named after validation functions, use those functions `name` property?